### ISSUE TITLE
List AccessibilityProvider should be mandatory

### ISSUE THREAD

**Issue by isidorn** [microsoft/vscode#92342]
List accessbiilityProvider should not be optional but mandatory. Currently clients of the list / tree forget to implement this and it is a core feature which must be supported always.
@jrieken also rasied this point

@joaomoreno we could make it mandatory in the `options` or it could be another paramter after the `renderers` in the list. Let me know what you think and prefer

---

⚠️  **Action Required**

You must now provide a `accessibilityProvider` to your list/tree. Please go through your instances and implement a meaningful one.

**Steps:** Checkout `joao/list-a11y` and compile. Remove the error in your file.

- [x] commentsTreeViewer.ts @rebornix 
- [x] extensionsViewer.ts @sandy081 
- [x] extensionsViews.ts @sandy081 
- [x] remote.ts @alexr00 
- [x] tunnelView.ts @alexr00

**Comment by joaomoreno** [comment]
Not going ahead with this. Even though this is true most of the cases, isn't not true in all of them. The new notebooks editor for example would have no user for this.

I

**Comment by joaomoreno** [comment]
Actually, we can make this at the Workbench widgets level only.

**Comment by joaomoreno** [comment]
Made notebook editor return `null`. cc @jrieken @rebornix

**Comment by joaomoreno** [comment]
Last one to fix this, please ping me.

**Comment by isidorn** [comment]
@joaomoreno I find it ugly that we have both `ariaProvider` and `accessibilityProvider`. Should those two be merged into one which has the mendatory `getAriaLabel`.
Code pointer https://github.com/microsoft/vscode/blob/404f66c13cb3c4ef0fb30db8a09b2967fc4b8367/src/vs/base/browser/ui/list/listWidget.ts#L846

**Comment by joaomoreno** [comment]
Yeah that is bad. Can you create a debt item for that? I'll merge them once this issue is closed.

**Comment by isidorn** [comment]
Thanks. Created https://github.com/microsoft/vscode/issues/94544

**Comment by sandy081** [comment]
@isidorn I cannot implement this for keybindings editor as I use `aria-labelledby` on the list element - 

https://github.com/microsoft/vscode/blob/dfb7a5ebd3ad07208974286ea13d88c87960fddf/src/vs/workbench/contrib/preferences/browser/keybindingsEditor.ts#L828-L830

Any suggestions?

**Comment by sandy081** [comment]
Cannot adopt in extensions using `PagedList` because it does not support the accessibility provider mapping the index with element

@joaomoreno FYI

**Comment by eamodio** [comment]
While starting to implement this for the Timeline view, I was thinking about adding an optional property to each `TimelineItem` for accessibility (maybe called `ariaLabel`?), because just having a default concatenation of  label and timestamp doesn't feel rich/accurate enough. Whereas a specific `TimelineProvider` can provide a richer, more accurate description for an item. For example, the Git `TimelineProvider` can provide a more meaningful combination of timestamp, author, commit message (maybe all of it rather than just the first line?), etc, e.g. _3 days ago, `<user>` committed `<commit-message>`_

Also while thinking about this, it seems that the same is true for custom tree views as well. Where a `TreeDataProvider` could provide a richer value than the current (which uses the tooltip if it exists and label if not). We of course could fallback to reasonable defaults when this new property doesn't exist.

Thoughts about adding a new property for both `TreeItem` and `TimelineItem`?

/cc @jrieken @isidorn @alexr00

**Comment by joaomoreno** [comment]
> Cannot adopt in extensions using `PagedList` because it does not support the accessibility provider mapping the index with element
> 
> @joaomoreno FYI

Sorry about this @sandy081, the `PagedList` had bad options typing. I've pushed a fix, can you try again?

**Comment by joaomoreno** [comment]
> @isidorn I cannot implement this for keybindings editor as I use `aria-labelledby` on the list element -
> 
> https://github.com/microsoft/vscode/blob/dfb7a5ebd3ad07208974286ea13d88c87960fddf/src/vs/workbench/contrib/preferences/browser/keybindingsEditor.ts#L828-L830
> 
> Any suggestions?

You shouldn't be accessing the parent DOM node, since you don't own it... If you need such a feature from the list/tree let me know.

**Comment by isidorn** [comment]
@joaomoreno I think @sandy081 does not put it on the list parent node, but on `monaco-list-row` element. At least that is what I see when I inspect the DOM.
@sandy081 do you have the reference to those child elements? Couldn't you just compute your `ariaLabel` by going to them and asking for their `ariaLabel` and concateneting.

**Comment by joaomoreno** [comment]
> @joaomoreno I think @sandy081 does not put it on the list parent node, but on `monaco-list-row` element. At least that is what I see when I inspect the DOM. The naming in code might be wrong.
> @sandy081 do you have the reference to those child elements? Couldn't you just compute your `ariaLabel` by going to them and asking for their `ariaLabel` and concateneting.

`monaco-list-row` is what I meant by parent. Users of the list do not own that element, they should never modify it. If `labeledby` is needed, it should be a feature of the accessibility provider.

**Comment by isidorn** [comment]
@joaomoreno yesm makes sense. So the `accessibilityProvider` needs to have one mandatory method `getAriaLabel`  and we could extend the return type that you can return an array of ids. Which would mean that the `labeledby` would be set.

Though let's wait for  @sandy081 and see if `labeledby` is really needed.

**Comment by sandy081** [comment]
Actually not. I can generate the `ariaLabel` directly and set it using `accessibilityProvider`.

**Comment by sandy081** [comment]
@joaomoreno 

> Sorry about this @sandy081, the PagedList had bad options typing. I've pushed a fix, can you try again?

I do not see any change. I am still called with index instead of extension element.

**Comment by joaomoreno** [comment]
@sandy081 You should do the work in the `joao/list-a11y` branch, like mentioned in the first post.

**Comment by sandy081** [comment]
Ah ok. Missed that. I checked in accessibility provider for keybindings editor into master. Do I have to revert that and move to this branch?

**Comment by joaomoreno** [comment]
No, you can leave that there.

**Comment by joaomoreno** [comment]
I've merged `master` into `joao/list-a11y`.

**Comment by alexr00** [comment]
Updated remote.ts and runnelView.ts.

**Comment by sandy081** [comment]
Done ✅

**Comment by rebornix** [comment]
Done ✅

**Comment by joaomoreno** [comment]
Thanks everyone
